### PR TITLE
refactor: cleanup duplicate functions and move apiversions to consts

### PR DESCRIPTION
**Pull request by CecileRobertMichon**
**Reason for Change**:

pre-refactor refactor: separating all non related changes in #2651 to make reviews easier and reverting possible if necessary.

**Issue Fixed**:

**Requirements**:

- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Notes**:

### CODE REVIEW

File: parts/masteroutputs.t [hunk 1]
```diff
@@ -6,7 +6,7 @@
       "value": ""
 {{end}}
     }
-{{if AnyAgentUsesAvailabilitySets}}
+{{if HasVMASAgentPool}}
```
Comment (CecileRobertMichon) [inline parts/masteroutputs.t]: renamed `AnyAgentUsesAvailabilitySets` to `HasVMASAgentPool` to match the VMSS equivalent

File: pkg/api/mocks.go [hunk 2]
```diff
@@ -232,3 +234,27 @@ func GetMockPropertiesWithCustomCloudProfile(name string, hasCustomCloudProfile,
 	}
 	return p
 }
+
+func getMockAddon(name string) KubernetesAddon {
```
Comment (CecileRobertMichon) [inline pkg/api/mocks.go]: moved this to mocks file (was in test file)

File: pkg/api/types.go [hunk 3]
```diff
@@ -925,18 +925,16 @@ func (p *Properties) GetAgentVMPrefix(a *AgentPoolProfile, index int) string {
 	return vmPrefix
 }
 
-// AnyAgentUsesVirtualMachineScaleSets checks whether any of the agents in the AgentPool use VMSS or not
-func (p *Properties) AnyAgentUsesVirtualMachineScaleSets() bool {
```
Comment (CecileRobertMichon) [inline pkg/api/types.go]: removing this function has it is a duplicate of `HasVMSSAgentPool`

File: pkg/api/types.go [hunk 4]
```diff
@@ -925,18 +925,16 @@ func (p *Properties) GetAgentVMPrefix(a *AgentPoolProfile, index int) string {
...
-	for _, agentProfile := range p.AgentPoolProfiles {
-		if agentProfile.IsVirtualMachineScaleSets() {
-			return true
-		}
+// GetVMType returns the type of VM "vmss" or "standard" to be passed to the cloud provider
+func (p *Properties) GetVMType() string {
```
Comment (CecileRobertMichon) [inline pkg/api/types.go]: new function to get VMType

File: pkg/api/types.go [hunk 5]
```diff
@@ -925,18 +925,16 @@ func (p *Properties) GetAgentVMPrefix(a *AgentPoolProfile, index int) string {
...
+	if p.HasVMSSAgentPool() {
+		return VMSSVMType
 	}
-	return false
+	return StandardVMType
 }
 
-// AnyAgentUsesAvailabilitySets checks whether any of the agents in the AgentPool use VMAS or not
-func (p *Properties) AnyAgentUsesAvailabilitySets() bool {
+// HasVMASAgentPool checks whether any of the agents in the AgentPool use VMAS or not
```
Comment (CecileRobertMichon) [inline pkg/api/types.go]: renamed `AnyAgentUsesAvailabilitySets` to `HasVMASAgentPool` to match the VMSS equivalent

File: pkg/engine/armvariables.go [hunk 6]
```diff
@@ -145,23 +144,22 @@ func getK8sMasterVars(cs *api.ContainerService) (map[string]interface{}, error)
...
 		"masterFqdnPrefix":              "[tolower(parameters('masterEndpointDNSNamePrefix'))]",
-		"apiVersionCompute":             "2019-07-01",
-		"apiVersionDeployments":         "2018-06-01",
-		"apiVersionStorage":             "2018-07-01",
-		"apiVersionKeyVault":            "2018-02-14",
-		"apiVersionNetwork":             "2018-08-01",
-		"apiVersionManagedIdentity":     "2015-08-31-preview",
-		"apiVersionAuthorizationUser":   "2018-09-01-preview",
-		"apiVersionAuthorizationSystem": "2018-01-01-preview",
+		"apiVersionCompute":             api.APIVersionCompute,
```
Comment (CecileRobertMichon) [inline pkg/engine/armvariables.go]: moved all apiVersions to constants

File: pkg/engine/armvariables.go [hunk 7]
```diff
@@ -398,6 +391,7 @@ func getK8sMasterVars(cs *api.ContainerService) (map[string]interface{}, error)
 	}
 
 	masterVars["nsgID"] = "[resourceId('Microsoft.Network/networkSecurityGroups',variables('nsgName'))]"
+	masterVars["subnetName"] = cs.Properties.GetSubnetName()
```
Comment (CecileRobertMichon) [inline pkg/engine/armvariables.go]: use `GetSubnetName()` function instead of embedding logic in armvariables file

File: pkg/api/types_test.go [hunk 8]
```diff
@@ -6571,97 +6555,7 @@ func TestAnyAgentUsesAvailabilitySets(t *testing.T) {
...
-			expected: true,
-		},
-	}
-
-	for _, test := range tests {
-		test := test
-		t.Run(test.name, func(t *testing.T) {
-			t.Parallel()
-			ret := test.p.AnyAgentUsesVirtualMachineScaleSets()
+			ret := test.p.HasVMASAgentPool()
```
Comment (jackfrancis) [inline pkg/api/types_test.go]: Wrong substitution?

File: pkg/engine/armvariables.go [hunk 9]
```diff
@@ -413,16 +407,16 @@ func getK8sMasterVars(cs *api.ContainerService) (map[string]interface{}, error)
 		masterVars["storageAccountBaseName"] = ""
 	}
 
-	if cs.Properties.AnyAgentUsesVirtualMachineScaleSets() {
+	masterVars["vmType"] = cs.Properties.GetVMType()
+
+	if cs.Properties.HasVMSSAgentPool() {
 		masterVars["primaryAvailabilitySetName"] = ""
-		masterVars["vmType"] = "vmss"
```
Comment (jackfrancis) [inline pkg/engine/armvariables.go]: yuck, thanks

File: pkg/api/types_test.go (hunk 8 above)
Comment (CecileRobertMichon) [reply pkg/api/types_test.go]: no it's just that I removed the whole test for the function I deleted so the diff appears weird. Test for AnyAgentUsesVirtualMachineScaleSets is completely gone, test for HasVMASAgentPool starts on line 6476 and continues here.

Comment (jackfrancis) [reply pkg/api/types_test.go]: or maybe the diff is confusing here

File: pkg/engine/armvariables.go (hunk 7 above)
Comment (jackfrancis) [reply pkg/engine/armvariables.go]: it's not clear to me that the existing GetSubnetName() returns the equivalent string responses as the above foo

Comment (CecileRobertMichon) [reply pkg/engine/armvariables.go]: yeah I think you're right, looking into it... I assumed the existing function did the same thing but looks like it didn't

Comment (jackfrancis) [reply pkg/engine/armvariables.go]: I think the existing implementation depends upon ARM params usage patterns, so maybe that means it's a candidate for a future refactor phase.

File: pkg/api/types.go [hunk 10]
```diff
@@ -1050,15 +1048,18 @@ func (p *Properties) GetVirtualNetworkName() string {
 // GetSubnetName returns the subnet name of the cluster based on its current configuration.
 func (p *Properties) GetSubnetName() string {
```
Comment (CecileRobertMichon) [inline pkg/api/types.go]: @jackfrancis I fixed the function to look more like the armvariables logic so it's easier to compare and we were indeed missing the condition `if !p.IsHostedMasterProfile() && !p.MasterProfile.IsCustomVNET() && p.MasterProfile.IsVirtualMachineScaleSets()` then `subnetName = "subnetmaster"` 

Hopefully the tests at https://github.com/Azure/aks-engine/pull/2659/files#diff-de120f39e8d8448b9aa430e66c997a97R4820 help bring more confidence that it actually does the same thing...

That being the naming logic doesn't really make sense and I don't see why vmss masters need a different subnet name but I'm guessing that's just the natural evolution of the two templates being in separate files previously, and changing the behavior would mean risking breaking upgrade/scale so I'm going to stay away from that for now.

Comment (CecileRobertMichon) [reply pkg/api/types.go]: I recommend looking at the file directly https://github.com/Azure/aks-engine/blob/01e8d7720d6483b95467294e2561cdebef260dd0/pkg/api/types.go#L1049, the diff is pretty hard to read...

File: pkg/engine/engine.go [hunk 11]
```diff
@@ -744,10 +744,7 @@ func getClusterAutoscalerAddonFuncMap(addon api.KubernetesAddon, cs *api.Contain
 			return api.GetClusterAutoscalerNodesConfig(addon, cs)
 		},
 		"GetVMType": func() string {
-			if cs.Properties.AnyAgentUsesVirtualMachineScaleSets() {
-				return base64.StdEncoding.EncodeToString([]byte("vmss"))
-			}
-			return base64.StdEncoding.EncodeToString([]byte("standard"))
```
Comment (CecileRobertMichon) [inline pkg/engine/engine.go]: @jackfrancis do you have any idea why we are base64 encoding the result here?

Comment (CecileRobertMichon) [reply pkg/engine/engine.go]: looks like it's being used as data in addon yaml specs, I renamed the function so it's clear

Comment (jackfrancis) [reply pkg/engine/engine.go]: Right. We should rename this to something like `GetVMTypeForClusterAutoscaler`

Comment (jackfrancis) [reply pkg/engine/engine.go]: `GetBase64EncodedVMType` also works :)

**Pull request merged**